File: modules/action/src/main/scala/action/sourcing/CommandConsumer.scala
package action.sourcing

import io.simplesource.kafka.model.CommandResponse
import model.topics
import model.topics.TopicNamer
import org.apache.kafka.streams.StreamsBuilder
import org.apache.kafka.streams.kstream.{Consumed, KStream}

object CommandConsumer {

  def commandResponseStream[A, I, K, C](spec: CommandSpec[A, I, K, C],
                                        commandTopicNamer: TopicNamer,
                                        builder: StreamsBuilder): KStream[K, CommandResponse] = {
    val responseByAggregate = builder
      .stream[K, CommandResponse](commandTopicNamer(topics.CommandTopic.response),
                                  Consumed.`with`(spec.serdes.aggregateKey, spec.serdes.commandResponse()))
      .peek(SourcingStream.logValues[K, CommandResponse]("commandResponseStream"))
    responseByAggregate
  }
}

File: modules/action/src/main/java/io/simplesource/action/sourcing/CommandConsumer.java
package io.simplesource.action.sourcing

import io.simplesource.kafka.model.CommandResponse
import shared.topics.TopicNamer;
import lombok.val;
import org.apache.kafka.streams.StreamsBuilder
import org.apache.kafka.streams.kstream.{Consumed, KStream}
import shared.topics.{TopicNamer, TopicTypes}

class CommandConsumer {

  static <A, I, K, C> KStream<K, CommandResponse> commandResponseStream(CommandSpec<A, I, K, C> spec,
                                                                        TopicNamer commandTopicNamer,
                                                                        StreamsBuilder builder) = {
    val responseByAggregate = builder
      .stream(commandTopicNamer(TopicTypes.CommandTopic.response),
                                  Consumed.`with`(spec.serdes.aggregateKey, spec.serdes.commandResponse()))
      .peek(SourcingStream.logValues[K, CommandResponse]("commandResponseStream"))
    responseByAggregate
  }
}
